neurodamus/core/stimuli.py

```diff
@@ -491,17 +493,88 @@ def attach_to(self, section, position=0.5):
         )
 
 
-# EStim class is a derivative of TStim for stimuli with an extracelular electrode. The main
-# difference is that it collects all elementary stimuli pulses and converts them using a
-# VirtualElectrode object before it injects anything
-#
-# The stimulus is defined on the hoc level by using the addpoint function for every (step) change
-# in extracellular electrode voltage. At this stage only step changes can be used. Gradual,
-# i.e. sinusoidal changes will be implemented in the future
-# After every step has been defined, you have to call initElec() to perform the frequency dependent
-# transformation. This transformation turns e_electrode into e_extracellular at distance d=1 micron
-# from the electrode. After the transformation is complete, NO MORE STEPS CAN BE ADDED!
-# You can then use inject() to first scale e_extracellular down by a distance dependent factor
-# and then vector.play() it into the currently accessed compartment
-#
-# TODO: 1. more stimulus primitives than step. 2. a dt of 0.1 ms is hardcoded. make this flexible!
+class ElectrodeSource(SignalSource):
+    """Constructs an extracellular potential field as the sum of multiple user-defined e-fields,
+    and applies the resulting signal to the segment's e_extracellular.
+
+    Args:
+        fields: list of user-defined electric field components (e.g. cosinuoid fields)
+        duration: duration of the signal, not including ramp up and ramp down.
+        ramp_up_time: duration during which the signal amplitude ramps up linearly to 0, in ms
+        ramp_down_time: duration during which the signal amplitude ramps down linearly to 0, in ms
+        base_position: coordinate [x, y, z] of the ground point where potential is 0
```

Comment:
nitpick: maybe we can point out that it is usually the soma baricenter if not specified? Is it the soma baricenter?

Reply:
Yes, I can add that in the comment. But there is no default value of base_position, user has to specify it.